src/Security/Security.php

```diff
@@ -596,180 +670,180 @@ public static function clearLoginMessage()
      * For multiple authenticators, Security_MultiAuthenticatorLogin is used.
      * See getTemplatesFor and getIncludeTemplate for how to override template logic
      *
-     * @return string|HTTPResponse Returns the "login" page as HTML code.
+     * @param $request
+     * @return HTTPResponse|string Returns the "login" page as HTML code.
+     * @throws HTTPResponse_Exception
      */
-    public function login()
+    public function login($request, $service = Authenticator::LOGIN)
```

Comment:
as suggested above, $request could be optional.

Also, there's a real issue with passing in $service here. This is a routing URL handler, and the request system can sometimes pass in additional arguments which may cause this method to break. (e.g. See FormRequestHandler, which passes in about 5 arguments to form actions).

```
return $controller->$funcName($vars, $this->form, $request, $this)
```

A safer idea is to assign a $url_handler to login that takes the service name as an argument. Then your login method looks as the below:

```
public function login($request = null) {
	if (!$request) {
		$request = $this->getRequest();
	}
	$service = $request ? $request->param('ID') : 'default';
}
```

You will need to address the issue with args being either flags (Authenticator::LOGIN) or names. I suggest just sticking with service names, since you probably won't want flags being passable from the HTTP request would you?

CMSSecurity::login() would be similar, but would default to 'cms' instead of 'default'